src/nl/hannahsten/texifyidea/refactoring/inlinecommand/LatexInlineCommandProcessor.kt

```diff
@@ -0,0 +1,124 @@
+import nl.hannahsten.texifyidea.psi.LatexRequiredParamContent
+import nl.hannahsten.texifyidea.util.childrenOfType
+import nl.hannahsten.texifyidea.util.definitionCommand
+
+/**
+ * Recieves settings and a target and performs the requested refactoring
+ *
+ * @see com.intellij.refactoring.inline.InlineMethodProcessor
+ *
+ * @author jojo2357
+ */
+class LatexInlineCommandProcessor(
+    myProject: Project,
+    private val inlineCommand: LatexCommands,
+    private val originalReference: PsiElement?,
+    private val isInlineThisOnly: Boolean,
+    private val isKeepTheDeclaration: Boolean,
+    private val myScope: SearchScope = GlobalSearchScope.projectScope(myProject)
+) : BaseRefactoringProcessor(myProject) {
+
+    override fun createUsageViewDescriptor(usages: Array<out UsageInfo>): UsageViewDescriptor {
+        return LatexInlineCommandDescriptor(inlineCommand)
+    }
+
+    override fun findUsages(): Array<UsageInfo> {
+        val tempreferences =
+            ReferencesSearch.search(inlineCommand.definitionCommand()!!, myScope).findAll().asSequence()
+
+        return tempreferences
+            .distinct()
+            .map(PsiReference::getElement)
+            .map(::UsageInfo)
+            .toList()
+            .toTypedArray()
+    }
+
+    override fun performRefactoring(usages: Array<out UsageInfo>) {
+        if (isInlineThisOnly) {
+            if (originalReference != null)
+                replaceUsage(originalReference)
+            else if (!ApplicationManager.getApplication().isUnitTestMode)
+                throw IllegalStateException("Inline this requested with no original reference (" + usages.size + ")")
+            // else we are probably trying to elicit no response because this is illegal
+        }
+        else if (usages.isNotEmpty()) {
+            ApplicationManager.getApplication().runWriteAction {
+                CommandProcessor.getInstance().executeCommand(myProject, {
+                    for (replaceUsage in usages) {
+                        val replacereference = replaceUsage.element ?: continue
+                        replaceUsage(replacereference)
+                    }
+                    if (!isKeepTheDeclaration)
+                        inlineCommand.delete()
+                }, commandName, "Texify")
+            }
+        }
+    }
+
+    /**
+     * Does the lifting by...replacing usages with the entirety of the file to be inlined
+     *
+     * @param psiElement The element to remove and replace with the contents of the file
+     */
+    private fun replaceUsage(psiElement: PsiElement) {
+        val calledRequiredArgs = psiElement.childrenOfType(LatexRequiredParamContent::class).toList()
+        val calledOptionalArgs = psiElement.childrenOfType(LatexKeyvalPair::class).toList()
+
+        val parentOptionalArgs = inlineCommand.childrenOfType(LatexKeyvalPair::class).toList()
+
+        if (calledOptionalArgs.size > 1)
+            return // error case
```

Comment:
In this case, I think the command definition would still be removed, but it would be better to not touch the document if there is a case we don't support?

Reply:
Rectified later with a return value now